### DISCUSSION TITLE
Proposal: self keyword

### DISCUSSION THREAD

**Discussion by p-lindberg** [dotnet/csharplang#2618] [Language Ideas]
I propose that the following:

```
public class Foo
{
    public Foo Bar()
    {
        return this;
    }
}
```

is given the following shorthand syntax:
```
public class Foo
{
    public self Bar()
    {
    }
}
```
In short, the self return type keyword makes a member method automatically return its own instance.

This would not just be syntactic sugar, but would also function as a guarantee that a method returning its own type does not in fact return another instance of the same type. This way, callers can be certain they are dealing with the same instance after calling such a method. This would be useful when building frameworks that employ method chaining.

Inheritance would need to be considered. I think the most natural course of action would be to let self-returning methods on derived types return a reference of the derived type, and not the base type. This would provide additional convenience in situations where you perform method chaining on derived objects: you would still be accessing an object of derived type after the first method invocation, which means that you don't have to downcast after calling a base method. Example:

```
public class Base
{
    public self Foo()
    {
    }
}

public class Derived : Base
{
    public self Bar()
    {
    }
}
```
This would be consumed as follows:
```
var derived = new Derived();
derived.Foo().Bar();
```
currently, this would look like:
```
public class Base
{
    public Base Foo()
    {
    }
}

public class Derived : Base
{
    public Derived Bar()
    {
    }
}
```
chaining these methods requires a cast:
```
var derived = new Derived();
((Derived)derived.Foo()).Bar();
```

**Comment by brabebhin** [comment]
Wouldn't this break encapsulation?

**Comment by p-lindberg** [comment]
I don't see how it would. Could you clarify?

**Comment by HaloFour** [comment]
Check out #252 and #902.  Although they're different in that they reuse the existing `this` reserved word rather than adding a new contextual keyword.
